File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-0.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        hash = [-1] * 256 

        left = 0
        right = 0
        max_len = 0

        while right < len(s):
            if hash[ord(s[right])] != -1:
                if hash[ord(s[right])] >= left:
                    left = hash[ord(s[right])] + 1
            
            curr_length = right - left + 1
            max_len = max(curr_length, max_len)

            hash[ord(s[right])] = right 
            right = right + 1

        return max_len        

File: Data Structures & Algorithms/eating-bananas/submission-0.ts
class Solution {
    /**
     * @param {number[]} piles
     * @param {number} h
     * @return {number}
     */

    findTotalHours(piles:number[], mid: number):number {
        let totalHours = 0

        for (let i = 0; i < piles.length; i++) {
            totalHours += Math.ceil(piles[i] / mid)
        }
        return totalHours
    }

    minEatingSpeed(piles: number[], h: number): number {
        let left = 1;
        let right = Math.max(...piles)

        let ans = -1

        while (left <= right) {
            let mid = Math.floor((left + right) / 2);
            
            let totalHours = this.findTotalHours(piles, mid)
            if (totalHours <= h) {
                ans = mid
                right = mid - 1
            } else {
                left = mid + 1
            }
        }
        return ans
    }
}
